### PR TITLE
Change the Gutenberg integration to use XCFramework

### PR DESCRIPTION
**Pull request by mokagio**
_This one was a long time coming... kept having to tweak the setup_ 😰

This PRs adds all the setup necessary to fetch and use Gutenberg as an XCFramework via commit or tag.

The setup is relatively straightforward. If `Gutenberg/version.rb` specifies a commit, the code that declares the `pod 'Gutenberg'` dependency will try and fetch the `podspec` for that commit from our CDN.

You can see the setup in action in https://github.com/wordpress-mobile/WordPress-iOS/pull/20792 and https://github.com/wordpress-mobile/WordPress-iOS/pull/20845. In #20845 in particular, notice how everything is green apart from a UI test which is 1) unrelated to the editor (insights) and 2) most likely failing due to async behavior.

image

Notice that this PR doesn't switch Gutenberg to a commit or tag. Assuming this PR will be merged before the next minor version of Gutenberg ships (meaning 1.98.0), the switch will happen when the Gutenberg folks update to that new version.

## Testing

You can try replicating the changes from #20792 locally, then run `bundle exec pod install`. Verify then that the editor works as expected. _@fluiddot  @geriux @SiobhyB @dcalhoun I'd appreciate your input as well as knowing if I wired up things correctly. Thanks!_

## Regression Notes

1. Potential unintended areas of impact – N.A.
2. What I did to test those areas of impact (or what existing automated tests I relied on) – N.A.
3. What automated tests I added (or what prevented me from doing so) – N.A.

---

**PR submission checklist:**

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes. **N.A.**
- [x] I have considered adding accessibility improvements for my changes. **N.A.**
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary. **N.A.**

**UI changes testing checklist:** Not a UI PR.

### CODE REVIEW

File: Gutenberg/Gutenberg.podspec
Comment (mokagio) [inline Gutenberg/Gutenberg.podspec]: Here's how it looks in action

Screenshot 2023-06-02 at 2 06 26 pm


Admittedly, that's not very visible. Plus, using the relative path makes it confusing to know where the file actually is. I plan to iterate on this, but I'm saving it for later because going back and forth on this setup often requires paying the download cost of a fresh gutenberg-mobile repo with submodules.

Once both commit and tag install are supported, we won't have to download that repo anymore and we'll be able to iterate faster on this.

File: .buildkite/cache-builder.yml
Comment (mokagio) [inline .buildkite/cache-builder.yml]: Admittedly, this was unrelated with the Gutenberg setup.

I did run into some Xcode version compatibility issue during the long development of this integration, but I think it should no longer be a problem.

Having said that, I think we might as well use the latest toolchain, no?

**Comment by geriux** [comment]
Hey there @mokagio 👋 

First of all, thank you for working on this 🚀 

### Tests

- ✅ Smoke tested the build from this PR and it works correctly, tested media uploads, story, pages, and post creation.
- ✅ Ran the app locally with the bundle that comes from Gutenberg Mobile.
- ⚠️ Running the app locally with Metro running doesn't work as it used to, before it used to load the local bundle provided by Metro but instead now is using the packed bundle. Maybe this is a side-effect of using Gutenberg as an XCFramework, there's code in Gutenberg to either provide the bundle file or the code provided locally with Metro ([see](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-bridge/ios/Gutenberg.swift#L263-L266)). **I wouldn't see this as a blocker** since we use the Gutenberg demo app as well for development, and if using the `LOCAL_GUTENBERG` flag allows Metro work (I couldn't test this, see the below).
- ❌ Running the app locally using the `LOCAL_GUTENBERG` flag didn't work I got a `Could not find Gutenberg bundle in either XCFramework or Pods.` build error. I do see the Gutenberg code in Xcode under the (Development Pods).

**Comment by mokagio** [comment]
Thanks for giving this a go @geriux ! 🙇‍♂️ 

> ❌ Running the app locally using the `LOCAL_GUTENBERG` flag didn't work I got a Could not find Gutenberg bundle in either XCFramework or Pods. build error. I do see the Gutenberg code in Xcode under the (Development Pods).

Interesting. The changes I made to [that build script](https://github.com/wordpress-mobile/WordPress-iOS/pull/20717/files#diff-ddc99ad6c4ca26dda4022e5d208bcc130c1a1677c8ea9472247adfc86505de5c) should only have added a search path, so I wonder how it worked before 🤔 

For the record, there is no record of the `App.js` file in the `Development Pods` version of `Gutenberg`

image

I'll look into it.

**Comment by geriux** [comment]
> For the record, there is no record of the `App.js` file in the `Development Pods` version of `Gutenberg`

I think that's expected since we use Metro during development so by just having metro running the app should pick up the local bundle through that.

**Comment by mokagio** [comment]
> so I wonder how it worked before 🤔

Update. Turns out this has "always" been an issue and was not diagnosed earlier because (I'm guessing) no one ever run Gutenberg in local mode without having first installed an upstream version.

This can be verified locally, on `trunk`, like so:

```bash
rm -rf Pods && \
  rm -rf DerivedData && \
  LOCAL_GUTENBERG=1 bundle exec pod install && \
  bundle exec fastlane build_wordpress_for_testing device:"iPhone SE (3rd generation)"
```

In `build_wordpress_for_testing`, the Gutenberg script will fail.

**Comment by mokagio** [comment]
@geriux 

> > For the record, there is no record of the App.js file in the Development Pods version of Gutenberg
>
> I think that's expected since we use Metro during development so by just having metro running the app should pick up the local bundle through that.

Gotcha. That's definitely something I should keep looking into.

At first, I thought it explained the behavior I reported in [my latest comment](https://github.com/wordpress-mobile/WordPress-iOS/pull/20717#issuecomment-1603761213) but thinking about it those are two different thing. The script failing to find the `js` file at build time is different from the app failing "to connect to Metro (right terminology?)" at runtime.

**Comment by geriux** [comment]
> Gotcha. That's definitely something I should keep looking into.

Thank you!

>  from the app failing "to connect to Metro (right terminology?)

That is correct! 😄

File: Gutenberg/cocoapods_helpers.rb
Comment (crazytonyli) [inline Gutenberg/cocoapods_helpers.rb]: What do you think merging the two if blocks into one?
```suggestion
  else
    pod 'Gutenberg', podspec: "https://cdn.a8c-ci.services/gutenberg-mobile/Gutenberg-#{options[:tag] || options[:commit]}.podspec"
  end
```

File: Gutenberg/cocoapods_helpers.rb
Comment (crazytonyli) [inline Gutenberg/cocoapods_helpers.rb]: Nitpick: moving this return statement to the first line to make unexpected cases more prominent: 

```ruby
# when referencing via a tag or commit, we download pre-built frameworks
return if options.key?(:tag) || options.key?(:commit)

podspec_prefix = options[:path]
raise 'Unexpected Gutenberg dependencies configuration' if podspec_prefix.nil?

podspec_prefix += '/third-party-podspecs'
...
```

File: Gutenberg/version.rb
Comment (crazytonyli) [inline Gutenberg/version.rb]: I'm wondering if we should manually build and upload xcframeworks for `v1.97.1` (maybe even including a few releases back too). For two reasons: it's not obvious what release this commit belongs and more importantly someone may want to use an older releases for whatever reason (like verifying a bug), but they won't be able to do that because the old release doesn't exist in the S3 bucket.

**Review by crazytonyli — CHANGES_REQUESTED** [review]
I'm marking my review as "Request changes" because I'd like to get your thoughts on building xcframeworks for current and a few older Gutenberge releases.

File: Gutenberg/cocoapods_helpers.rb
Comment (mokagio) [reply Gutenberg/cocoapods_helpers.rb]: Good point. Built on top of this suggestion in 1d8deb1b6cce8a5faa4af976d42506b454147d3f to also add an error if neither is set.

File: Gutenberg/cocoapods_helpers.rb
Comment (mokagio) [reply Gutenberg/cocoapods_helpers.rb]: Done in 90ea9004ba35da1acfc8c4d96c5cbba68860aae8

File: Gutenberg/version.rb
Comment (mokagio) [reply Gutenberg/version.rb]: > I'm wondering if we should manually build and upload xcframeworks for v1.97.1 (maybe even including a few releases back too

This can be done. Sure 👍 

However, the build for v1.98.0 is available, so there is no blocking concern to merging this. If that's okay with you, I'll rebase on `trunk` (and loose the "changes since you last saw this") so that we can use that build. Otherwise, I can put make a merge commit.

Comment (crazytonyli) [reply Gutenberg/version.rb]: I'm okay with either way. Whatever is convenient for you would work for me 😃

Comment (mokagio) [reply Gutenberg/version.rb]: I rebased. 

Unfortunately, as CI will show, since be bumped the iOS deployment target from 14 to 15, Gutenberg 1.98.0 no longer works:

image

So, we're back to square one in regards to having to manually publish new versions. 

In parallel, there's work going on updating React Native on the Gutenberg end, which may be best suited as the version where to make the switch 😞 

- https://github.com/wordpress-mobile/gutenberg-mobile/pull/5908
- https://github.com/wordpress-mobile/gutenberg-mobile/pull/5912

Comment (fluiddot) [reply Gutenberg/version.rb]: > In parallel, there's work going on updating React Native on the Gutenberg end, which may be best suited as the version where to make the switch 😞
> 
> * [Workaround RNReanimated `pod install` failure gutenberg-mobile#5908](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5908)
> * [Workaround React Native 0.71 and XCFramework setup issues (without Hermes) gutenberg-](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5912)

Since the RN upgrade is blocked due to issues with the Reanimated Pod, which would be solved using the XCFramework. Maybe we could link this work with the ongoing work of the upgrade. I.e. the RN upgrade and the XCFramwork integration would land at the same time.

We could wrap up the workaround ([Workaround React Native 0.71 and XCFramework setup issues (without Hermes)](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5912/files#top)) and merge it into the [Gutenberg Mobile PR](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5881). And then, we could rebase this branch with the changes needed for the upgrade in WP-iOS ([PR reference](https://github.com/wordpress-mobile/WordPress-iOS/pull/20956)), including updating the Gutenberg Mobile reference. WDYT?

Comment (mokagio) [reply Gutenberg/version.rb]: Sounds good to me 👍

Comment (fluiddot) [reply Gutenberg/version.rb]: @mokagio I'm trying to test the workaround for solving the issue with Reanimated, but when installing the pods in WP-iOS I'm getting the error shared [here](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5908).

```
Error: Cannot find module 'react-native/package.json'
Require stack:
- /Users/fluiddot/workspace/WordPress-iOS/[eval]
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:931:15)
    at Function.resolve (internal/modules/cjs/helpers.js:113:19)
    at [eval]:1:9
    at Script.runInThisContext (vm.js:134:12)
    at Object.runInThisContext (vm.js:310:38)
    at internal/process/execution.js:81:19
    at [eval]-wrapper:6:22
    at evalScript (internal/process/execution.js:80:60)
    at internal/main/eval_string.js:27:3 {
  code: 'MODULE_NOT_FOUND',
  requireStack: [ '/Users/fluiddot/workspace/WordPress-iOS/[eval]' ]
}

[!] Invalid `RNReanimated.podspec` file: no implicit conversion of nil into String.

 #  from /var/folders/s7/lwl9q12j2_31hw8y1vkmln180000gn/T/d20230630-7085-146eqk6/RNReanimated.podspec:5
 #  -------------------------------------------
 #  reanimated_package_json = JSON.parse(File.read(File.join(__dir__, "package.json")))
 >  config = find_config()
 #  assert_no_multiple_instances(config)
 #  -------------------------------------------
```

I'm wondering if we'd need to specify as done [here](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5912/files#diff-141dafd8df7f3d127b6295567b663372841c0bc9f142fcf433b6858cc3f95cf3R57) that Reanimated should be fetched from the forked repository. WDYT?

On a similar note, I wanted to ask if now, that we integrate Gutenberg using the XCFramework, we still need to reference and install the Pod dependencies of Gutenberg.

https://github.com/wordpress-mobile/WordPress-iOS/blob/0f8cc000b479d2ad10fea632155c186f4c597ee1/Gutenberg/cocoapods_helpers.rb#L14-L60

Thanks you very much for the help on this 🙇 !

Comment (mokagio) [reply Gutenberg/version.rb]: Hey @fluiddot 👋 

Is the error occurring on this branch? A standard `bundle exec pod install` should not result in that error ([as can be seen in CI](https://buildkite.com/automattic/wordpress-ios/builds/15153#01891a4f-cc58-4d21-8967-1e6d4d2b0bed)) because of the XCFramework.

The only way I can think for that error happening is when running using a local version of Gutenberg, `LOCAL_GUTENBERG=true bundle exec pod install`?

I'll take a look at that flow now. FWIW, that seems consistent with what @geriux [reported](https://github.com/wordpress-mobile/WordPress-iOS/pull/20717#issuecomment-1603142170) during earlier testing.

Comment (fluiddot) [reply Gutenberg/version.rb]: > Is the error occurring on this branch? A standard `bundle exec pod install` should not result in that error ([as can be seen in CI](https://buildkite.com/automattic/wordpress-ios/builds/15153#01891a4f-cc58-4d21-8967-1e6d4d2b0bed)) because of the XCFramework.
>
> The only way I can think for that error happening is when running using a local version of Gutenberg, `LOCAL_GUTENBERG=true bundle exec pod install`?

@mokagio Yes, but as you pointed out, I encountered this error when using the local version of Gutenberg. It works If I run `bundle exec pod install` using the tag `v1.98.0` as the Gutenberg Mobile reference 👍 .

> I'll take a look at that flow now. FWIW, that seems consistent with what @geriux [reported](https://github.com/wordpress-mobile/WordPress-iOS/pull/20717#issuecomment-1603142170) during earlier testing.

That would be great, thanks 🙇 ! We rely on using the local version of Gutenberg when testing and debugging changes we need to make on the native side related to the editor. Hence, being able to keep using it would be very helpful for the project. Let me know if I can help anyhow on this part.

**Comment by fluiddot** [comment]
Heads up that I updated [the PR related to the RN upgrade 0.71.11](https://github.com/wordpress-mobile/WordPress-iOS/pull/20956) with these changes. I'm happy to share that [the installable builds were created](https://github.com/wordpress-mobile/WordPress-iOS/pull/20956#issuecomment-1622071488) 🎊.

There are only two minor changes on [that PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/20956) compared to this one:
* [Update RNReanimated pod source to wp-fork-2.17.0](https://github.com/wordpress-mobile/WordPress-iOS/pull/20956/commits/e35b419fdb5ddfea5ccec914f3d044c9a8782453)
* [Update Gutenberg Mobile reference](https://github.com/wordpress-mobile/WordPress-iOS/pull/20956/commits/31a4f102e372bd318e0f08e1c5937ef11c310265)

**Comment by fluiddot** [comment]
@mokagio Heads up that I created [this PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/21021) to fix the current failures we have in CI jobs. Let me know if you could take a look, thanks 🙇 !

**Comment by mokagio** [comment]
Thanks @fluiddot!

I received your message but I might run out of time to look into it today. I have a lot of PRs to review for the upcoming 22.8 code freeze.

**Comment by fluiddot** [comment]
> Thanks @fluiddot!
> 
> I received your message but I might run out of time to look into it today. I have a lot of PRs to review for the upcoming 22.8 code freeze.

No worries at all @mokagio. We can revisit this next week if you have more availability. Thanks!

**UPDATE:** I've just realized you will be away next week, so we can postpone this to the week after the next one.

**Comment by fluiddot** [comment]
@mokagio Heads up that I've updated the PR with `trunk` to get recent changes. From my side the PR is ready but I understand that we'd need first to land:
- https://github.com/wordpress-mobile/WordPress-iOS/pull/21021
- Fixes introduced in https://github.com/wordpress-mobile/WordPress-iOS/pull/21128

Let me know if I'm missing anything to set the XCFramework integration ready.

The UI tests are still failing even when using the latest changes introduced in the editor. We should keep investigating how to address them.

Regarding the merge, I'd like to note that since we need changes from the React Native upgrade, we should wait until those changes are merged. Alternatively, I'm updating https://github.com/wordpress-mobile/WordPress-iOS/pull/20956 with all changes needed for the RN upgrade. Maybe it's easier to merge them through that PR.

**Comment by fluiddot** [comment]
Closing in favor of https://github.com/wordpress-mobile/WordPress-iOS/pull/20956, as these changes need to be merged along with the React Native upgrade `0.71.11`.

Heads up that I'll re-assign current reviewers to that PR to validate the XCFramework integration. Thanks for your help 🙇 !

**Pull request closed without merging**
